### PR TITLE
[93] Fix DnD Constraints into different Constraint compartments

### PR DESCRIPTION
**Pull request by jerome-obeo**
Bug: https://github.com/eclipse-syson/syson/issues/93

### CODE REVIEW

File: backend/views/syson-diagram-common-view/src/main/java/org/eclipse/syson/diagram/common/view/services/ViewToolService.java [hunk 1]
```diff
@@ -268,4 +273,53 @@ public Element dropRequirementSubjectFromDiagram(Element droppedElement, Node dr
         }
         return droppedElement;
     }
+
+    public Element dropElementFromDiagramInRequirementAssumeConstraintCompartment(Element droppedElement, Node droppedNode, Element targetElement, Node targetNode, IEditingContext editingContext, IDiagramContext diagramContext,
+            Map<org.eclipse.sirius.components.view.diagram.NodeDescription, NodeDescription> convertedNodes) {
+        if (droppedElement instanceof ConstraintUsage droppedConstraint) {
+            if (targetElement instanceof RequirementUsage requirementUsage) {
```
Comment (AxelRICHARD) [inline backend/views/syson-diagram-common-view/src/main/java/org/eclipse/syson/diagram/common/view/services/ViewToolService.java:280]: Could you merge the two `if` in one please?

File: backend/views/syson-diagram-common-view/src/main/java/org/eclipse/syson/diagram/common/view/services/ViewToolService.java [hunk 2]
```diff
@@ -268,4 +273,53 @@ public Element dropRequirementSubjectFromDiagram(Element droppedElement, Node dr
...
+                diagramContext.getViewDeletionRequests().add(ViewDeletionRequest.newViewDeletionRequest().elementId(droppedNode.getId()).build());
+            }
+        }
+        return droppedElement;
+    }
+
+    public Element dropElementFromDiagramInRequirementRequireConstraintCompartment(Element droppedElement, Node droppedNode, Element targetElement, Node targetNode, IEditingContext editingContext, IDiagramContext diagramContext,
+            Map<org.eclipse.sirius.components.view.diagram.NodeDescription, NodeDescription> convertedNodes) {
+        if (droppedElement instanceof ConstraintUsage droppedConstraint) {
+            if (targetElement instanceof RequirementUsage requirementUsage) {
```
Comment (AxelRICHARD) [inline backend/views/syson-diagram-common-view/src/main/java/org/eclipse/syson/diagram/common/view/services/ViewToolService.java:292]: Could you let mix the two `if` in one please?

File: backend/views/syson-diagram-common-view/src/main/java/org/eclipse/syson/diagram/common/view/services/ViewToolService.java [hunk 3]
```diff
@@ -268,4 +273,53 @@ public Element dropRequirementSubjectFromDiagram(Element droppedElement, Node dr
...
+                this.moveContraintInRequirementConstraintCompartment(droppedConstraint, requirementUsage, RequirementConstraintKind.REQUIREMENT);
+                this.createView(droppedElement, editingContext, diagramContext, targetNode, convertedNodes, NodeContainmentKind.CHILD_NODE);
+                diagramContext.getViewDeletionRequests().add(ViewDeletionRequest.newViewDeletionRequest().elementId(droppedNode.getId()).build());
+            }
+        }
+        return droppedElement;
+    }
+
+    private void moveContraintInRequirementConstraintCompartment(ConstraintUsage droppedConstraint, RequirementUsage requirementUsage, RequirementConstraintKind kind) {
+        var oldMembership = droppedConstraint.eContainer();
```
Comment (AxelRICHARD) [inline backend/views/syson-diagram-common-view/src/main/java/org/eclipse/syson/diagram/common/view/services/ViewToolService.java:302]: Why create a new membership and delete the existing one?
Could you just move the existing membership instead?

File: backend/views/syson-diagram-common-view/src/main/java/org/eclipse/syson/diagram/common/view/services/ViewToolService.java [hunk 4]
```diff
@@ -268,4 +273,53 @@ public Element dropRequirementSubjectFromDiagram(Element droppedElement, Node dr
...
+        requirementUsage.getOwnedRelationship().add(membership);
+        EcoreUtil.delete(oldMembership);
+    }
+
+    public Element dropElementFromDiagramInConstraintCompartment(Element droppedElement, Node droppedNode, Element targetElement, Node targetNode, IEditingContext editingContext, IDiagramContext diagramContext,
+            Map<org.eclipse.sirius.components.view.diagram.NodeDescription, NodeDescription> convertedNodes) {
+        if (droppedElement instanceof ConstraintUsage droppedConstraint) {
+            if (targetElement instanceof ConstraintUsage || targetElement instanceof ConstraintDefinition) {
+                var oldMembership = droppedConstraint.eContainer();
+                var membership = SysmlFactory.eINSTANCE.createFeatureMembership();
```
Comment (AxelRICHARD) [inline backend/views/syson-diagram-common-view/src/main/java/org/eclipse/syson/diagram/common/view/services/ViewToolService.java:315]: Why create a new membership and delete the existing one?
Could you just move the existing membership instead?

File: backend/views/syson-diagram-common-view/src/main/java/org/eclipse/syson/diagram/common/view/services/ViewToolService.java (hunk 3 above)
Comment (AxelRICHARD) [reply backend/views/syson-diagram-common-view/src/main/java/org/eclipse/syson/diagram/common/view/services/ViewToolService.java:302]: Ok thank you for the clarification

File: backend/views/syson-diagram-common-view/src/main/java/org/eclipse/syson/diagram/common/view/services/ViewToolService.java (hunk 4 above)
Comment (jerome-obeo) [reply backend/views/syson-diagram-common-view/src/main/java/org/eclipse/syson/diagram/common/view/services/ViewToolService.java:315]: It's the same explanation as before.

File: backend/views/syson-diagram-common-view/src/main/java/org/eclipse/syson/diagram/common/view/services/ViewToolService.java (hunk 1 above)
Comment (jerome-obeo) [reply backend/views/syson-diagram-common-view/src/main/java/org/eclipse/syson/diagram/common/view/services/ViewToolService.java:280]: Done.

File: backend/views/syson-diagram-common-view/src/main/java/org/eclipse/syson/diagram/common/view/services/ViewToolService.java (hunk 2 above)
Comment (jerome-obeo) [reply backend/views/syson-diagram-common-view/src/main/java/org/eclipse/syson/diagram/common/view/services/ViewToolService.java:292]: Done.

**Pull request merged**